pkg/monitortests/kubeapiserver/legacykubeapiservermonitortests/early_e2e_disruption.go

```diff
@@ -0,0 +1,92 @@
+package legacykubeapiservermonitortests
+
+import (
+	"fmt"
+	"regexp"
+	"strings"
+	"time"
+
+	"github.com/openshift/origin/pkg/monitor/monitorapi"
+	"github.com/openshift/origin/pkg/test/ginkgo/junitapi"
+)
+
+// amount of time after E2E tests start that we consider "early"
+const preE2ECheckDuration = 3 * time.Minute
+const earlyE2ECheckDuration = 3 * time.Minute
+
+func testEarlyE2EAPIServerDisruption(events monitorapi.Intervals) []*junitapi.JUnitTestCase {
+	var e2eStart time.Time // anchor the time period we're interested in disruption
+	for _, event := range events {
+		if event.Source == monitorapi.SourceE2ETest {
+			e2eStart = event.From
+			break
+		}
+	}
+
+	const preTestName = "[Jira:\"oauth-apiserver\"] oauth servers should not experience disruption shortly before the start of E2E testing"
+	const earlyTestName = "[Jira:\"kube-apiserver\"] API servers should not experience disruption during the start of E2E testing"
```

Comment:
Apologies I'm misunderstood your plan earlier and the result I'm unsure about. I thought you would have both tests for *-api servers, one for before e2e testing, one for during or shortly after start of e2e testing. 

In current form, it's weird to see a test specifically for oauth backends *before* e2e testing testing, then a generic test for Jira kube-apiserver that actually checks kube/openshift+oauth, *during or shortly after* the start of e2e testing. For example, if all oauth goes down at the moment e2e is testing, i would like that as visible by itself as the separate oauth test, if we're really hunting for that symptom. As it stands that would be embedded in a more generic test and not in a specific oauth test, which it should be detected for the same issue. 

So I think (I'm open to suggestions) I would suggest two tests, one for oauth api, one for kube+openshift api, encompassing all of shortly before, during, or shortly after e2e testing started. Kube and openshift are probably closely related enough to lump together and call Jira kube-apiserver. Alternatively I'd be open to a third test to separate openshift-api, but realistically if kube goes openshift goes with it, it's rare we're hunting for a problem specific to openshift-api. 

Essentially: 

`
	const oauthTestName = "[Jira:\"oauth-apiserver\"] oauth API servers should not experience disruption near the start of E2E testing"
	const kubeTestName = "[Jira:\"kube-apiserver\"] kube API servers should not experience disruption near the start of E2E testing"
`

With minimal test adjustments to accomplish this. That feels more clear and consistent to me. We won't see a difference between disruption 2m before vs during e2e start vs 2min after unless we analyze test results, but that feels logically more like what we want to identify as if I'm hunting those bugs, it's not relevant to me, I only care that it happened.

What do you and Forrest think?

Reply:
With that approach we don't differentiate before from early in the start but we can get a better idea if oauth occurs at a higher rate near the start.  I'll +1 that.

Reply:
OK, so the junits now just separate oauth from apiserver. Junits now for "https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/test-platform-results/logs/periodic-ci-openshift-release-master-nightly-4.17-e2e-metal-ipi-ovn-kube-apiserver-rollout/1827763939853733888/ look like

> [
>  
> {"XMLName":{"Space":"","Local":""},"Name":"[Jira:\"kube-apiserver\"] kube API servers should not experience disruption near the start of E2E testing","Classname":"","Duration":0,"SkipMessage":null,"FailureOutput":{"XMLName":{"Space":"","Local":""},"Message":"","Output":
> "found 18 apiserver disruption events near E2E test start (Aug 25 15:22:37) with messages:
> Aug 25 15:26:02.000 - 1s    E backend-disruption-name/kube-api-new-connections connection/new disruption/openshift-tests reason/DisruptionBegan request-audit-id/70f3ad85-4e5c-4d9a-8b80-06b29869681a backend-disruption-name/kube-api-new-connections connection/new disruption/openshift-tests stopped responding to GET requests over new connections: Get \"https://api.ostest.test.metalkube.org:6443/api/v1/namespaces/default\": net/http: timeout awaiting response headers
> [...]
> Aug 25 15:26:27.000 - 2s    E backend-disruption-name/kube-api-reused-connections connection/reused disruption/openshift-tests reason/DisruptionBegan request-audit-id/c7f1a76b-ac29-4b9c-9f2f-3dfcabb587a9 backend-disruption-name/kube-api-reused-connections connection/reused disruption/openshift-tests stopped responding to GET requests over reused connections: Get \"https://api.ostest.test.metalkube.org:6443/api/v1/namespaces/default\": net/http: timeout awaiting response headers"
> },"SystemOut":"","SystemErr":""},
>  
> {"XMLName":{"Space":"","Local":""},"Name":"[Jira:\"kube-apiserver\"] kube API servers should not experience disruption near the start of E2E testing","Classname":"","Duration":0,"SkipMessage":null,"FailureOutput":null,"SystemOut":"","SystemErr":""},
>  
> {"XMLName":{"Space":"","Local":""},"Name":"[Jira:\"oauth-apiserver\"] oauth API servers should not experience disruption near the start of E2E testing","Classname":"","Duration":0,"SkipMessage":null,"FailureOutput":{"XMLName":{"Space":"","Local":""},"Message":"",
> "Output":"found 21 oauthserver disruption events near E2E test start (Aug 25 15:22:37) with messages:                                Aug 25 15:25:50.000 - 2s    E backend-disruption-name/oauth-api-reused-connections connection/reused disruption/openshift-tests reason/DisruptionBegan request-audit-id/0bdb665a-e1db-498d-bf33-777e28350326 backend-disruption-name/oauth-api-reused-connections connection/reused disruption/openshift-tests stopped responding to GET requests over reused connections: error running request: 503 Service Unavailable: service unavailable
> [...]
> Aug 25 15:26:27.000 - 1s    E backend-disruption-name/oauth-api-reused-connections connection/reused disruption/openshift-tests reason/DisruptionBegan request-audit-id/7934370d-3caf-4177-aab9-263cba9c9df6 backend-disruption-name/oauth-api-reused-connections connection/reused disruption/openshift-tests stopped responding to GET requests over reused connections: Get \"https://api.ostest.test.metalkube.org:6443/apis/oauth.openshift.io/v1/oauthclients\": net/http: timeout awaiting response headers"
> },"SystemOut":"","SystemErr":""},
>  
> {"XMLName":{"Space":"","Local":""},"Name":"[Jira:\"oauth-apiserver\"] oauth API servers should not experience disruption near the start of E2E testing","Classname":"","Duration":0,"SkipMessage":null,"FailureOutput":null,"SystemOut":"","SystemErr":""}
> ]
>